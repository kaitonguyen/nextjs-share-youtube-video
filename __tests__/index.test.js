// __tests__/index.test.jsx

import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import Navbar from "../components/navbar";
import Home from "../pages/index";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";

describe("Navbar", () => {
  it("renders login/register form if user do not logged in", () => {
    render(<Navbar email={""} />);

    const loginForm = screen.queryByTestId("login-form");

    expect(loginForm).toBeInTheDocument();
  });

  it("renders authenticated action buttons if user do not logged in", () => {
    render(<Navbar email={""} />);

    const actionButtons = screen.queryByTestId("authenticated-actions");

    expect(actionButtons).toBeNull();
  });

  it("no renders login/register form if user logged in", () => {
    render(<Navbar email={"nguyenhieuky@gmail.com"} />);

    const loginForm = screen.queryByTestId("login-form");

    expect(loginForm).toBeNull();
  });

  it("renders authenticated action buttons if user logged in", () => {
    render(<Navbar email={"nguyenhieuky@gmail.com"} />);

    const actionButtons = screen.queryByTestId("authenticated-actions");

    expect(actionButtons).toBeInTheDocument();
  });
});

describe("Login/Register", () => {
  it("cannot login if email and password are empty", () => {
    render(<Navbar email={""} />);
    const loginForm = screen.getByTestId("login-form");

    expect(loginForm).not.toBeValid();
  });

  it("cannot login if email is empty", () => {
    render(<Navbar email={""} />);

    const password = screen.getByPlaceholderText("password");
    const loginButton = screen.getByRole("login-button");
    const loginForm = screen.getByTestId("login-form");

    fireEvent.change(password, { target: { value: "password" } });

    expect(loginForm).not.toBeValid();
  });

  it("cannot login if email is invalid", async () => {
    render(<Navbar email={""} />);

    const email = screen.getByPlaceholderText("email");
    const password = screen.getByPlaceholderText("password");
    const loginForm = screen.getByTestId("login-form");

    fireEvent.change(password, { target: { value: "password" } });
    fireEvent.change(email, { target: { value: "" } });
    expect(loginForm).not.toBeValid();

    fireEvent.change(email, { target: { value: "email" } });
    expect(loginForm).not.toBeValid();

    fireEvent.change(email, { target: { value: "email@" } });
    expect(loginForm).not.toBeValid();

    fireEvent.change(email, { target: { value: " " } });
    expect(loginForm).not.toBeValid();

    fireEvent.change(email, { target: { value: "@" } });
    expect(loginForm).not.toBeValid();

    fireEvent.change(email, { target: { value: "@gmail" } });
    expect(loginForm).not.toBeValid();

    fireEvent.change(email, { target: { value: "@gmail.com" } });
    expect(loginForm).not.toBeValid();

    fireEvent.change(email, { target: { value: "@.com" } });
    expect(loginForm).not.toBeValid();

    fireEvent.change(email, { target: { value: "email@.com" } });
    expect(loginForm).not.toBeValid();

    fireEvent.change(email, { target: { value: "email@gmail" } });
    const mockLogin = jest.fn((email) => {
      return Promise.resolve({ email });
    });
    expect(mockLogin).not.toBeCalled();

    fireEvent.change(email, { target: { value: "email@gmail.com" } });
    expect(loginForm).toBeValid();
  });
});

describe("Logout", () => {
  it("renders login/register form", () => {
    const navbar = render(<Navbar email={"nguyenhieuky@gmail.com"} />);

    const logoutButton = screen.getByRole("logout-button");
    userEvent.click(logoutButton)
    expect(navbar).toMatchSnapshot();
  });
});

describe("Home - main content", () => {
  it("renders videos list", () => {
    render(<Home />);

    const videosList = screen.queryByTestId("videos-list");

    expect(videosList).toBeInTheDocument();
  });
});
