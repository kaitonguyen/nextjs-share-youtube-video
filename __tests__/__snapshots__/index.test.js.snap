// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logout renders login/register form 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="container max-w-6xl m-auto border-b border-b-stone-600 py-4 flex justify-between items-center"
      >
        <div
          class=""
        >
          <h1
            class="flex text-3xl font-bold"
            id="logo"
          >
            <a
              href="/"
            >
              <svg
                class="h-8 w-8 inline-block mr-2 mb-2"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M10.707 2.293a1 1 0 00-1.414 0l-7 7a1 1 0 001.414 1.414L4 10.414V17a1 1 0 001 1h2a1 1 0 001-1v-2a1 1 0 011-1h2a1 1 0 011 1v2a1 1 0 001 1h2a1 1 0 001-1v-6.586l.293.293a1 1 0 001.414-1.414l-7-7z"
                />
              </svg>
              Funny Movies
            </a>
          </h1>
        </div>
        <div
          class=""
        >
          <div
            class="flex md:gap-5 items-center"
            data-testid="authenticated-actions"
          >
            <span
              role="welcome-text"
            >
              Welcome 
              nguyenhieuky@gmail.com
            </span>
            <a
              class="px-5 py-3 bg-sky-600 text-white"
              href="/share"
            >
              Share a movie
            </a>
            <button
              class="px-5 py-3 bg-rose-500 text-white"
              role="logout-button"
            >
              Logout
            </button>
          </div>
        </div>
      </header>
    </div>
  </body>,
  "container": <div>
    <header
      class="container max-w-6xl m-auto border-b border-b-stone-600 py-4 flex justify-between items-center"
    >
      <div
        class=""
      >
        <h1
          class="flex text-3xl font-bold"
          id="logo"
        >
          <a
            href="/"
          >
            <svg
              class="h-8 w-8 inline-block mr-2 mb-2"
              fill="currentColor"
              viewBox="0 0 20 20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10.707 2.293a1 1 0 00-1.414 0l-7 7a1 1 0 001.414 1.414L4 10.414V17a1 1 0 001 1h2a1 1 0 001-1v-2a1 1 0 011-1h2a1 1 0 011 1v2a1 1 0 001 1h2a1 1 0 001-1v-6.586l.293.293a1 1 0 001.414-1.414l-7-7z"
              />
            </svg>
            Funny Movies
          </a>
        </h1>
      </div>
      <div
        class=""
      >
        <div
          class="flex md:gap-5 items-center"
          data-testid="authenticated-actions"
        >
          <span
            role="welcome-text"
          >
            Welcome 
            nguyenhieuky@gmail.com
          </span>
          <a
            class="px-5 py-3 bg-sky-600 text-white"
            href="/share"
          >
            Share a movie
          </a>
          <button
            class="px-5 py-3 bg-rose-500 text-white"
            role="logout-button"
          >
            Logout
          </button>
        </div>
      </div>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
